tutorials/data-science-preprocessing/index.md

```diff
@@ -0,0 +1,437 @@
+appropriate types:
+
+[embedmd]:# (src/preprocessing/clean.py /def convert_types/ /return.*/)
+```py
+def convert_types(record):
+    """Converts string values to their appropriate type."""
+    # Only the year part of the datetime string is significant
+    record['year'] = int(record['year'][:4])
+
+    record['mass'] = float(record['mass']) if 'mass' in record else None
+
+    geolocation = record['geolocation']
+    geolocation['latitude'] = float(geolocation['latitude'])
+    geolocation['longitude'] = float(geolocation['longitude'])
+
+    return record
+```
+
+After having completed the preprocessing step and gone on to explore the data,
+you may notice that there are some suspicious values. For example, some entries
+have the value `0.0` for both `latitude` and `longitude`. Your explorations in
+the [exploratory section](bigquery.md) of the process will inform the filters
+you'll need to clean the data up - in this case, by discarding invalid values,
+and making educated guesses for the correct value when you can:
+
+[embedmd]:# (src/preprocessing/clean.py /def filter_suspicious/ /yield.*/)
+```py
+def filter_suspicious(record):
+    """Filters records with suspicious values."""
+    if record['year'] < 100:
+        raise StopIteration()
+    if 100 <= record['year'] < 1000:
+        # Probably just forgot the leading '1'
+        record['year'] += 1000
+
+    geolocation = record['geolocation']
+    if not (geolocation['latitude'] or geolocation['longitude']):
+        raise StopIteration()
+
+    yield record
+```
+
+Finally, clean up some redundant fields, and flatten the record in preparation
+for insertion into a queryable database:
+
+[embedmd]:# (src/preprocessing/clean.py /def massage_rec/ /return.*/)
+```py
+def massage_rec(record):
+    """Massage keys with the 'rec' prefix."""
+    # These are redundant with the geolocation
+    del record['reclat']
+    del record['reclong']
+
+    # Flatten the geolocation into the object
+    record.update(record['geolocation'])
+    del record['geolocation']
+    # Probably don't need this. It seems to always be false
+    del record['needs_recoding']
+
+    # Remove the 'rec' prefix
+    record['class'] = record['recclass']
+    del record['recclass']  # Is it just me or is this pronounced "reckless"?
+
+    return record
+```
+
+## Define a Dataflow pipeline
+
+[Google Cloud Dataflow](/dataflow) uses the [Apache Beam SDK][beam]
+to define a processing pipeline for the data to go through. In this case, the
+data needs to be processed by each of these functions in succession and then
+inserted into [Google BigQuery](/bigquery), after being read from its
+original raw format.
+
+### Create a `Source` for JSON objects
+
+Apache Beam provides certain `Source` objects that can read entries from a file
+and emit them, but unfortunately does not provide one for json objects. So
+you'll need to define one. From [the documentation][filebasedsource], you must
```

Comment:
I think "From the documentation" is a little confusing here. Do you mean "As shown in this documentation"? It might be easiest to omit the opening clause and just say "You must define...". Then you could link the class name to its reference doc.